source/compiler/sc5.c

```diff
@@ -423,3 +450,173 @@ int pc_geterrorwarnings()
+ */
+static int minimum(int a,int b,int c)
+{
+  int min=a;
+  if(b<min)
+    min=b;
+  if(c<min)
+    min=c;
+  return min;
+}
+
+static int levenshtein_distance(const char *s,const char*t)
+{
+  //Step 1
+  int k,i,j,cost,*d,distance;
+  int n=strlen(s);
+  int m=strlen(t);
+  assert(n>0 && m>0);
+  d=(int*)malloc((sizeof(int))*(m+1)*(n+1));
+  m++;
+  n++;
+  //Step 2
+  for (k=0;k<n;k++)
+    d[k]=k;
+  for (k=0;k<m;k++)
+    d[k*n]=k;
+  //Step 3 and 4
+  for (i=1;i<n;i++) {
+    for (j=1;j<m;j++) {
+      //Step 5
+      cost= (tolower(s[i-1])!=tolower(t[j-1]));
+      //Step 6
+      d[j*n+i]=minimum(d[(j-1)*n+i]+1,d[j*n+i-1]+1,d[(j-1)*n+i-1]+cost);
+    } /* for */
+  } /* for */
+  distance=d[n*m-1];
+  free(d);
+  return distance;
+}
+
+static int find_closestsymbol_table(const char *name,const symbol *root,int symboltype,symbol **closestsym)
+{
+  int dist,closestdist=INT_MAX;
+  char symname[2*sNAMEMAX+16];
+  symbol *sym=root->next;
+  int ident,critdist;
+
+  assert(closestsym!=NULL);
+  *closestsym=NULL;
+  assert(name!=NULL);
+  critdist=strlen(name)/2;  /* for short names, allow only a single edit */
+  if (critdist>MAX_EDIT_DIST)
+    critdist=MAX_EDIT_DIST;
+  while (sym!=NULL) {
+    funcdisplayname(symname,sym->name);
+    ident=sym->ident;
+    if (symboltype==iARRAY && ident==iREFARRAY)
+      ident=iARRAY;     /* reference arrays match arrays */
+    else if (symboltype==iVARIABLE && (sym->ident==iCONSTEXPR || sym->ident==iREFERENCE || sym->ident==iARRAY || sym->ident==iREFARRAY))
+      ident=iVARIABLE;  /* when requesting variables, constants are also ok */
+    if ((symboltype==ident || (symboltype==iVARIABLE && ident==iFUNCTN)) && ((ident!=iFUNCTN && ident!=iLABEL) || (sym->usage & uDEFINE)!=0)) {
+      dist=levenshtein_distance(name,symname);
+      if (dist<closestdist && dist<=critdist) {
+        *closestsym=sym;
+        closestdist=dist;
+      } /* if */
+    } /* if */
+    sym=sym->next;
+  } /* while */
+  return closestdist;
+}
+
+static symbol *find_closestsymbol(const char *name,int symboltype)
+{
+  symbol *symloc,*symglb;
+  int distloc,distglb;
+
+  if (sc_status==statFIRST)
+    return NULL;
+  assert(name!=NULL);
+  if (strlen(name)==0)
+    return NULL;
+  distloc=find_closestsymbol_table(name,&loctab,symboltype,&symloc);
+  if (distloc<=1)
+    distglb=INT_MAX; /* don't bother searching in the global table */
+  else
+    distglb=find_closestsymbol_table(name,&glbtab,symboltype,&symglb);
+  return (distglb<distloc) ? symglb : symloc;
+}
+
+static constvalue *find_closestautomaton(const char *name)
+{
+  constvalue *ptr=sc_automaton_tab.first;
+  constvalue *closestmatch=NULL;
+  int dist,closestdist=INT_MAX;
+
+  assert(name!=NULL);
+  while (ptr!=NULL) {
+    if (strlen(ptr->name)>0) {
+      dist=levenshtein_distance(name,ptr->name);
+      if (dist<closestdist && dist<=MAX_EDIT_DIST) {
+        closestmatch=ptr;
+        closestdist=dist;
+        if (dist<=1)
```

Comment:
There's a lot of repetitive code here, which is never great, but I do see why - there are very subte differences in things like the loop condition, for example `find_closeststate` checking the automata is correct so as not to suggest matching states from other ones.

One difference I'm not clear on is this `break`.  It looks like a nice optimisation - if there is a very similar symbol, suggest that and don't bother to continue searching.  However, it is only in this one function, not all of them (which is, of course, the danger with repeated code).

Reply:
Yeah, looks like I forgot to add that check in the other functions. Thanks for noting this.